c2_utils/include/mfx_gralloc_instance.h

```diff
@@ -0,0 +1,38 @@
+// Copyright (c) 2017-2021 Intel Corporation
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in all
+// copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+// SOFTWARE.
+
+#pragma once
+#include <mutex>
+#include "mfx_gralloc_interface.h"
+#include "mfx_gralloc1.h"
+#include "mfx_gralloc4.h"
+
+class MfxGrallocInstance
+{
+public:
+    MfxGrallocInstance() = delete;
+    MfxGrallocInstance(const MfxGrallocInstance&) = delete;
+    MfxGrallocInstance& operator=(const MfxGrallocInstance&) = delete;
+
+    static std::shared_ptr<IMfxGrallocModule> getInstance();
+private:
+    static std::mutex m_mutex;
+    static std::shared_ptr<IMfxGrallocModule> m_instance;
```

Comment:
Can we have this s_instance as public member variable? And it is initialized when we know it is using VIDEO MEMORY in mfx_c2_decoder_components.cpp or mfx_c2_encoder_components.cpp. And use it as MfxGrallocInstance::s_instance in other files. Is it possible?

Reply:
You mean m_instance? This pointer will only be initialized when the first call the function getInstance(). The most safest and direct way to obtain m_instance is to use getInstance(). This is the most standard way to use singleton pattern.

Reply:
What I mean is to declare this instance member variable as public, like this, and initialize this instance once we got to know the memory type is MFX_IOPATTERN_OUT_VIDEO_MEMORY. 

class MfxGrallocInstance
{
public:
    MfxGrallocInstance() = delete;
    MfxGrallocInstance(const MfxGrallocInstance&) = delete;
    MfxGrallocInstance& operator=(const MfxGrallocInstance&) = delete;

    static std::shared_ptr<IMfxGrallocModule> getInstance();
public:
    static std::mutex m_mutex;
    static std::shared_ptr<IMfxGrallocModule> m_instance;
}

In line https://github.com/projectceladon/MediaSDK_C2/blob/master/c2_components/src/mfx_c2_decoder_component.cpp#L1039, we can do it like this,
if (m_mfxVideoParams.IOPattern == MFX_IOPATTERN_OUT_VIDEO_MEMORY) {
    MfxGrallocInstance::m_instance = MfxGrallocInstance::getInstance();
    if (MfxGrallocInstance::m_instance == nullptr) {
        res = MFX_ERR_NULL_PTR;
    }
}

After this initialization, we can use m_instance directly. Is it OK?

Reply:
I don't think this is good idea, it is not different with before. And you also need to call getInstance() to initialize the pointer in different module. Because you can't guarantee which is the first call. You will get the pointer m_instance directly when you call getInstance() and m_instance has been initialized. And it is the thread safe.

Reply:
We just call getInstance() only once for each decoder/encoder. This initialization happened when we create MfxC2DecoderComponent/MfxC2EncoderComponent. You can refer https://github.com/projectceladon/MediaSDK_C2/blob/master/c2_components/src/mfx_c2_decoder_component.cpp#L681. If m_instrance is nullptr, we will fail to create the component.